src/engine/Union.cpp

```diff
@@ -347,172 +346,250 @@ struct Wrapper {
   const IdTable& idTable_;
   const LocalVocab& localVocab_;
 };
-Result::IdTableVocabPair& moveOrCopy(Result::IdTableVocabPair& element) {
-  return element;
+Result::IdTableVocabPair moveOrCopy(Result::IdTableVocabPair& element) {
+  return std::move(element);
 }
 Result::IdTableVocabPair moveOrCopy(const Wrapper& element) {
   return {element.idTable_.clone(), element.localVocab_.clone()};
 }
-}  // namespace
 
-// _____________________________________________________________________________
-// Always inline makes makes a huge difference on large datasets.
-template <size_t SPAN_SIZE>
-AD_ALWAYS_INLINE bool Union::isSmaller(const auto& row1,
-                                       const auto& row2) const {
-  using StaticRange = std::span<const ColumnIndex, SPAN_SIZE>;
-  for (auto& col : StaticRange{targetOrder_}) {
-    ColumnIndex index1 = _columnOrigins.at(col).at(0);
-    ColumnIndex index2 = _columnOrigins.at(col).at(1);
-    if (index1 == NO_COLUMN) {
-      return true;
-    }
-    if (index2 == NO_COLUMN) {
-      return false;
+template <size_t SPAN_SIZE, typename Range1, typename Range2, typename Func>
+struct SortedUnionImpl
+    : ad_utility::InputRangeFromGet<Result::IdTableVocabPair> {
+  std::shared_ptr<const Result> result1_;
+  std::shared_ptr<const Result> result2_;
+  Range1 range1_;
+  std::optional<typename Range1::iterator> it1_ = std::nullopt;
+  size_t index1_ = 0;
```

Comment:
As this struct is now rather long, and rather decoupled,
Can you move it to a separate file + add some comments ++ maybe even split declarations and definitions, such that it becomes a little clearer.